feat: Add gatt read/write async response.

zhongzhijie1 opened the pull request:
depends-on: [open-vela/external_zblue/pull/107]

zhongzhijie1 committed:
GATTS: Support deferred write/read responses

bug: v/78372

Root cause: Receive and app reply run in different threads, require async handling.
Use 32-bit request_id encoding handle and opcode for framework service compatibility.

Signed-off-by: zhongzhijie1 <zhongzhijie1@xiaomi.com>

zhongzhijie1 committed:
Manager GATT server test: Normalize read-only/read-write/pts char access.

bug: v/78372

move PTS test char to tools/manager (replace SAL workaround); place Manager char value at tool (app) level.

Signed-off-by: zhongzhijie1 <zhongzhijie1@xiaomi.com>

zhongzhijie1 commented:
still under discuss in zephyr community

https://github.com/zephyrproject-rtos/zephyr/pull/97594#issuecomment-3544803967

liuX10 commented on tools/gatt_server.c:
```diff
@@ -118,18 +125,64 @@ uint16_t tx_char_ccc_changed(void* srv_handle, bt_address_t* addr, uint16_t attr
+{
+    gatts_device_t* device;
+    uint16_t payload_len;
+    uint8_t rsp_data[MAX_ATTRIBUTE_SIZE];
```
use heap memory.

zhongzhijie1 replied:
malloc and then free? okay

liuX10 commented on tools/gatt_server.c:
```diff
@@ -107,7 +112,9 @@ enum {
 uint8_t read_pts_char_value[] = { 'H', 'e', 'l', 'l', 'o', ' ', 'P', 'T', 'S', '!' };
-uint8_t read_char_value[] = { 'H', 'e', 'l', 'l', 'o', ' ', 'V', 'E', 'L', 'A', '!' };
+uint8_t read_only_char_value[] = { 'H', 'e', 'l', 'l', 'o', ' ', 'V', 'E', 'L', 'A', '!' };
+uint8_t read_write_char_value[MAX_ATTRIBUTE_SIZE] = { 'H', 'e', 'l', 'l', 'o', ' ', 'V', 'E', 'L', 'A', '!' };
```
we must need an array[512] at .data segment?

zhongzhijie1 replied:
this can let remote gatt client write some value, and then read to verify, I remember there was such a case in PTS.

zhongzhijie1 committed:
Remove redundant gatt server local read/write implement.

bug: v/78372

cause of open-vela/external_zblue/pull/107, now use async way for gatts
read/write response, so no need local way.

Signed-off-by: zhongzhijie1 <zhongzhijie1@xiaomi.com>

gzh-terry commented on service/stacks/zephyr/sal_gatt_server_interface.c:
```diff
@@ -154,49 +152,53 @@ static struct bt_gatt_attr server_db[CONFIG_GATT_SERVER_MAX_ATTRIBUTES];
 
-    if_gatts_on_received_element_write_request(&addr, GATT_OPS_WRITE_REQUEST, element->handle, (uint8_t*)buf, offset, len);
+    if (flags & GATT_WRITE_FLAGS_CMD_PREPARE_EXECUTE) {
+        request_id = REQUEST_ID_INVALID;
```
Is this an error? Should we return `-EINVAL` directly?

zhongzhijie1 replied:
now use REQUEST_ID_NORSP instead of REQUEST_ID_INVALID,

currenty can not handle prepare write and execute write, so return BT_GATT_ERR(BT_ATT_ERR_UNLIKELY);

zhongzhijie1 commented:
just refresh clang format, no file change
